### DISCUSSION TITLE
Ajuda: Erro na Distribuição de DF-e Tipo 3: consChNFe

### DISCUSSION THREAD

**Discussion by herbertsmatheus** [TadaSoftware/PyNFe#214] [Q&A] [answered]
Olá,

Eu gostaria de obter o xml de notas fiscais através da chave de acesso. Tentei a consulta tipo 3 que achei nesse [wiki](https://github.com/TadaSoftware/PyNFe/wiki/Consulta-Distribui%C3%A7%C3%A3o), porém ao rodar o código o xml gerado contém a seguinte mensagem:

_Server Error in '/' Application.
The resource cannot be found. 
Description: HTTP 404. The resource you are looking for (or one of its dependencies) could have been removed, had its name changed, or is temporarily unavailable.  Please review the following URL and make sure that it is spelled correctly. 
Requested URL: /NFeDistribuicaoDFe/NFeDistribuicaoDFe.asmx_

Como posso resolver esse problema?
Obrigado!

**Comment by leogregianin** [comment, marked as answer]
@herbertsmatheus a branch `main` do repositório foi atualizado com uma correção na URL de homologação da DistribuicaoDFe. Pode refazer seus testes?

**Reply by herbertsmatheus** [reply]
Bom dia Leo,

Muito obrigado, creio que o problema foi resolvido, mas agora retornou um outro erro:

`<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
<soap:Body>
<nfeDistDFeInteresseResponse xmlns="http://www.portalfiscal.inf.br/nfe/wsdl/NFeDistribuicaoDFe">
<nfeDistDFeInteresseResult>
<retDistDFeInt xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.portalfiscal.inf.br/nfe" versao="1.01">
<tpAmb>2</tpAmb>
<verAplic>1.6.0</verAplic>
<cStat>215</cStat>
<xMotivo>Rejeicao: Falha no esquema xml</xMotivo>
<dhResp>2022-09-06T10:26:43-03:00</dhResp>
<ultNSU>000000000000000</ultNSU>
<maxNSU>000000000000000</maxNSU>
</retDistDFeInt>
</nfeDistDFeInteresseResult>
</nfeDistDFeInteresseResponse>
</soap:Body>
</soap:Envelope>`

**Reply by leogregianin** [reply]
Somente com esse `envelope soap` do retorno não dá para saber exatamente o que pode ser. Como o motivo que consta no XML foi `<xMotivo>Rejeicao: Falha no esquema xml</xMotivo>`, imagino que seja os dados inseridos de CNPJ e Chave de Acesso que devem ser somente números, sem espaços ou outros tipos de caracteres.

**Reply by herbertsmatheus** [reply]
Será que pode ser pq no meu caso é CPF ao invés de CNPJ?

Segue abaixo meu código:

`from pynfe.processamento.comunicacao import ComunicacaoSefaz

certificado = '/Users/matheusherberts/Downloads/cert.pfx'
senha = '1234'
uf = 'mt'
homologacao = True

CNPJ = '25043340959'
CHAVE = '51220702003402007935550180000003331007230909'

con = ComunicacaoSefaz(uf, certificado, senha, homologacao)
xml = con.consulta_distribuicao(cnpj=CNPJ, chave=CHAVE)

with open('response.xml', 'w') as f:
    f.write(xml.text)`
    
   Tentei trocar o arg cnpj por cpf, mas aí o erro é que a NF-e não existe para a chave informada.

**Reply by leogregianin** [reply]
Para CNPJ seria:

```python
CNPJ = '12345678901234'
xml = con.consulta_distribuicao(cnpj=CNPJ, chave=CHAVE)
```

Para CPF seria: 
```python
CPF = '12345678901'
xml = con.consulta_distribuicao(cpf=CPF, chave=CHAVE)
```

Ou, de uma forma genérica:

```python
documento = 'cpf ou cnpj'
if len(documento) == 14:
   xml = con.consulta_distribuicao(cnpj=documento, chave=CHAVE)
else:
   xml = con.consulta_distribuicao(cpf=documento, chave=CHAVE) 
```

**Reply by herbertsmatheus** [reply]
fiz com dessa maneira para o cpf (cpf=CPF), porém agora obtenho a seguinte resposta:

`<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
<soap:Body>
<nfeDistDFeInteresseResponse xmlns="http://www.portalfiscal.inf.br/nfe/wsdl/NFeDistribuicaoDFe">
<nfeDistDFeInteresseResult>
<retDistDFeInt xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.portalfiscal.inf.br/nfe" versao="1.01">
<tpAmb>2</tpAmb>
<verAplic>1.6.0</verAplic>
<cStat>217</cStat>
<xMotivo>
Rejeicao: NF-e inexistente para a chave de acesso informada
</xMotivo>
<dhResp>2022-09-06T14:44:49-03:00</dhResp>
</retDistDFeInt>
</nfeDistDFeInteresseResult>
</nfeDistDFeInteresseResponse>
</soap:Body>
</soap:Envelope>`

**Reply by herbertsmatheus** [reply]
Não sei se isso poderia afetar, mas no caso o CPF que uso possui 2 cadastros no sefaz, pq temos duas inscrições estaduais.
Na questão de chave de acesso creio que isso não seria o problema visto que a mesma funciona no site da receita.

**Reply by herbertsmatheus** [reply]
aqui está o código

`from pynfe.processamento.comunicacao import ComunicacaoSefaz

certificado = '/Users/matheusherberts/Downloads/cert.pfx'
senha = '1234'
uf = 'MT'
homologacao = True

CPF = '25043340959'
CHAVE = '51220800135201000104550050002463971621162976'

con = ComunicacaoSefaz(uf, certificado, senha, homologacao)
xml = con.consulta_distribuicao(cpf=CPF, chave=CHAVE)

with open('response.xml', 'w') as f:
    f.write(xml.text)`

**Reply by leogregianin** [reply]
Inscrições Estaduais não tem nenhum influência nisso. Você informou uma chave de acesso que está em ambiente de produção e está usando `homologacao = True` ?

**Reply by herbertsmatheus** [reply]
ahhh, não sabia disso. Valeu!
Agora deu certo, só que o xml que obtive não era bem o que eu esperava.

Como eu faço para obter o xml em si?

**Reply by leogregianin** [reply]
Você deve ter recebido o resumo do XML. Para obter o XML completo é necessário fazer a Manifestação do Destinatário, tem um exemplo [aqui](https://github.com/TadaSoftware/PyNFe/wiki/Manifesta%C3%A7%C3%A3o-Destinat%C3%A1rio) e depois fazer a consulta novamente. Deve ter uma demora da Receita Federal entre essas etapas.

E para automatizar todo o fluxo de notas recebidas, tem esse exemplo [aqui](https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao-por-NSU), ou seja, consulta a sequencia de numeração de NSU, faz a manifestação e consulta novamente.

**Reply by herbertsmatheus** [reply]
Era bem isso.

Tentei fazer o manifesto da nota só que obtive um erro bem similar ao primeiro da conversa:

`<!DOCTYPE html>
<html>
    <head>
        <title>The resource cannot be found.</title>
        <meta name="viewport" content="width=device-width" />
        <style>
         body {font-family:"Verdana";font-weight:normal;font-size: .7em;color:black;} 
         p {font-family:"Verdana";font-weight:normal;color:black;margin-top: -5px}
         b {font-family:"Verdana";font-weight:bold;color:black;margin-top: -5px}
         H1 { font-family:"Verdana";font-weight:normal;font-size:18pt;color:red }
         H2 { font-family:"Verdana";font-weight:normal;font-size:14pt;color:maroon }
         pre {font-family:"Consolas","Lucida Console",Monospace;font-size:11pt;margin:0;padding:0.5em;line-height:14pt}
         .marker {font-weight: bold; color: black;text-decoration: none;}
         .version {color: gray;}
         .error {margin-bottom: 10px;}
         .expandable { text-decoration:underline; font-weight:bold; color:navy; cursor:pointer; }
         @media screen and (max-width: 639px) {
          pre { width: 440px; overflow: auto; white-space: pre-wrap; word-wrap: break-word; }
         }
         @media screen and (max-width: 479px) {
          pre { width: 280px; }
         }
        </style>
    </head>

    <body bgcolor="white">

            <span><H1>Server Error in '/' Application.<hr width=100% size=1 color=silver></H1>

            <h2> <i>The resource cannot be found.</i> </h2></span>

            <font face="Arial, Helvetica, Geneva, SunSans-Regular, sans-serif ">

            <b> Description: </b>HTTP 404. The resource you are looking for (or one of its dependencies) could have been removed, had its name changed, or is temporarily unavailable. &nbsp;Please review the following URL and make sure that it is spelled correctly.
            <br><br>

            <b> Requested URL: </b>/NFeRecepcaoEvento4/NFeRecepcaoEvento4.asmx<br><br>

            </font>

    </body>
</html>`
